clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp

```diff
@@ -1000,7 +1006,18 @@ class CIRAllocaLowering
                   rewriter.getIntegerAttr(rewriter.getIndexType(), 1));
     auto elementTy = getTypeConverter()->convertType(op.getAllocaType());
     auto resultTy = getTypeConverter()->convertType(op.getResult().getType());
-    // TODO: Verification between the CIR alloca AS and the one from data layout
+    // Verification between the CIR alloca AS and the one from data layout.
+    {
+      auto resPtrTy = mlir::cast<mlir::LLVM::LLVMPointerType>(resultTy);
+      auto dlAllocaASAttr = mlir::cast_if_present<mlir::IntegerAttr>(
+          dataLayout.getAllocaMemorySpace());
+      // Absence means 0
+      // TODO: The query for the alloca AS should be done through CIRDataLayout
+      // instead to reuse the logic of interpret null attr as 0.
+      auto dlAllocaAS = dlAllocaASAttr ? dlAllocaASAttr.getInt() : 0;
+      assert(dlAllocaAS == resPtrTy.getAddressSpace() &&
```

Comment:
Can we return a failure instead? Luxury we don't have in CIRGen.

Reply:
Thanks for the reminder! Converted to `op.emitError()`.